src/bare-metal/microcontrollers/debugging.md

```diff
@@ -0,0 +1,38 @@
+# Debugging
+
+Embed.toml:
+
+```toml
+[default.general]
+chip = "nrf52833_xxAA"
+
+[debug.gdb]
+enabled = true
+```
+
+In one terminal:
+
+```sh
+cargo embed --bin board_support debug
+```
+
+In another terminal:
+
+```sh
+gdb-multiarch target/thumbv7em-none-eabihf/debug/board_support --eval-command="target remote :1337"
+```
+
+<details>
+
+In GDB, try running:
+
+```gdb
+b src/bin/board_support.rs:28
+b src/bin/board_support.rs:29
+b src/bin/board_support.rs:31
+c
+c
+c
```

Comment:
Could you describe what the expected output is? I tried it and got

```
% gdb-multiarch target/thumbv7em-none-eabihf/debug/board_support --eval-command="target remote :1337"

GNU gdb (Debian 12.1-4) 12.1
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
target/thumbv7em-none-eabihf/debug/board_support: No such file or directory.
Remote debugging using :1337
warning: No executable has been specified and target does not support
determining executable automatically.  Try using the "file" command.
0x00000100 in ?? ()
(gdb) b src/bin/board_support.rs:28
No symbol table is loaded.  Use the "file" command.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (src/bin/board_support.rs:28) pending.
(gdb) b src/bin/board_support.rs:29
No symbol table is loaded.  Use the "file" command.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 2 (src/bin/board_support.rs:29) pending.
(gdb) b src/bin/board_support.rs:31
No symbol table is loaded.  Use the "file" command.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 3 (src/bin/board_support.rs:31) pending.
(gdb) c
Continuing.
c
c
c


(gdb) 
```

The light turned on, but that's it :smile: The many warnings about the lack of a symbol table were a bit confusing.

Reply:
Okay, it works better when I run it in the same directory:

```
Reading symbols from target/thumbv7em-none-eabihf/debug/board_support...
Remote debugging using :1337
microbit_common::display::nonblocking::control::{impl#0}::display_row_leds (
    self=<error reading variable: Cannot access memory at address 0x2002006c>, 
    row=<error reading variable: Cannot access memory at address 0x20020070>, 
    cols=<error reading variable: Cannot access memory at address 0x20020074>)
    at src/display/nonblocking/control.rs:134
134	            p0.outset.write(|w| w.bits(rows_to_set | cols_to_set));
(gdb) b src/bin/board_support.rs:28
Breakpoint 1 at 0x18e: file src/bin/board_support.rs, line 29.
Note: automatically using hardware breakpoints for read-only addresses.
(gdb) b src/bin/board_support.rs:29
Note: breakpoint 1 also set at pc 0x18e.
Breakpoint 2 at 0x18e: file src/bin/board_support.rs, line 29.
(gdb) b src/bin/board_support.rs:31
Breakpoint 3 at 0x1b8: file src/bin/board_support.rs, line 32.
(gdb) c
Continuing.

Breakpoint 1, board_support::__cortex_m_rt_main () at src/bin/board_support.rs:29
29	    board.display_pins.col1.set_low().unwrap();
(gdb) c
Continuing.

Program received signal SIGINT, Interrupt.
nrf_hal_common::gpio::p0::{impl#144}::set_low<nrf_hal_common::gpio::PushPull> (self=0x2001fff4) at /usr/local/google/home/mgeisler/.cargo/registry/src/github.com-1ecc6299db9ec823/nrf-hal-common-0.14.1/src/gpio.rs:581
581	                    fn set_low(&mut self) -> Result<(), Self::Error> {
(gdb) c
Continuing.

Breakpoint 3, board_support::__cortex_m_rt_main () at src/bin/board_support.rs:32
32	    loop {}
(gdb) 
```

However, the first two breakpoints seem to be duplicated? The `Program received signal SIGINT, Interrupt.` also looks scary?

Reply:
Hmm, looks like the line numbers are wrong, I'll send a PR to fix that. I'm not sure what the message about `SIGINT` is from, there's no such thing as `SIGINT` here. I have also seen it halt at lines other than the breakpoints I set, and I'm not sure why. I think it's still worth demonstrating debugging though, even if it's a bit buggy.